docs/ADR-8-Use-RIO-in-cardano‐cli.md

```diff
@@ -0,0 +1,73 @@
+# Status
+- [ ] Adopted YYYY/MM/DD
+
+# Context
+- In `cardano-cli` We have the (anti-)pattern `ExceptT someError IO someResult` 292 times in our codebase.
+![image](https://github.com/user-attachments/assets/f405dfca-7d75-404c-b291-3d5140fc8093)
+
+- The vast majority of these errors are not used to recover, they are propagated and reported to the user.
+```
+main :: IO ()
+main = toplevelExceptionHandler $ do
+  Crypto.cryptoInit
+
+  envCli <- getEnvCli
+
+  GHC.mkTextEncoding "UTF-8" >>= GHC.setLocaleEncoding
+#ifdef UNIX
+  _ <- setFileCreationMask (otherModes `unionFileModes` groupModes)
+#endif
+  co <- Opt.customExecParser pref (opts envCli)
+
+  orDie (docToText . renderClientCommandError) $ runClientCommand co
+...
+...
+...
+
+runClientCommand :: ClientCommand -> ExceptT ClientCommandErrors IO ()
+runClientCommand = \case
+  AnyEraCommand cmds ->
+    firstExceptT (CmdError (renderAnyEraCommand cmds)) $ runAnyEraCommand cmds
+  AddressCommand cmds ->
+    firstExceptT AddressCmdError $ runAddressCmds cmds
+  NodeCommands cmds ->
+    runNodeCmds cmds
+      & firstExceptT NodeCmdError
+  ByronCommand cmds ->
+    firstExceptT ByronClientError $ runByronClientCommand cmds
+  CompatibleCommands cmd ->
+    firstExceptT (BackwardCompatibleError (renderAnyCompatibleCommand cmd)) $
+      runAnyCompatibleCommand cmd
+  HashCmds cmds ->
+    firstExceptT HashCmdError $ runHashCmds cmds
+  KeyCommands cmds ->
+    firstExceptT KeyCmdError $ runKeyCmds cmds
+  LegacyCmds cmds ->
+    firstExceptT (CmdError (renderLegacyCommand cmds)) $ runLegacyCmds cmds
+  QueryCommands cmds ->
+    firstExceptT QueryCmdError $ runQueryCmds cmds
+  CliPingCommand cmds ->
+    firstExceptT PingClientError $ runPingCmd cmds
+  CliDebugCmds cmds ->
+    firstExceptT DebugCmdError $ runDebugCmds cmds
+  Help pprefs allParserInfo ->
+    runHelp pprefs allParserInfo
+  DisplayVersion ->
+    runDisplayVersion
+```
+- As a result we have a lot of errors wrapped in errors which makes the code unwieldly and difficult to compose with other code blocks. See image below of incidences of poor composability where we use `firstExceptT` (and sometimes `first`) to wrap errors in other errors.
+![image](https://github.com/user-attachments/assets/6a50e8f2-9140-4b7e-afa5-ff778d696742)
+ 
+- The ExceptT IO is a known anti-pattern for these reasons and others as per:
+  - https://github.com/haskell-effectful/effectful/blob/master/transformers.md
+  - https://tech.fpcomplete.com/blog/2016/11/exceptions-best-practices-haskell/
+
+
+# Decision
+- Error handling: All errors will be converted to exceptions that will be caught by a single exception handler at the top level.
+- Top level monad: [RIO](https://hackage.haskell.org/package/rio-0.1.22.0/docs/RIO.html#t:RIO). See the [ReaderT design pattern](https://tech.fpcomplete.com/blog/2017/06/readert-design-pattern/).
```

Comment:
1. There's no description of a proposed solution. What will be the convention for the  definition exception types? You didn't provide any, so I guess, the goal will be to define an exception near the usage site without [building exceptions hierarchy](https://hackage.haskell.org/package/base-4.21.0.0/docs/Control-Exception.html#t:Exception)?
We should be at least be using `Pretty` for rendering of exceptions, to leverage already existing instances like `Pretty Lovelace` - this needs to be inlcuded in this ADR

2. Why there's suddenly a mention of `ReaderT` design pattern? I don't see how it's related to exception handling.

Reply:
Proposed exception handling here: https://github.com/IntersectMBO/cardano-cli/pull/1033/commits/989ba8a544277e73548947eb74404f25ddf76d17 

I'll add it to the ADR.

Reply:
Could you address the second point, please?